commerce/bookstore/schema.tql

```diff
@@ -0,0 +1,134 @@
+#
+# Copyright (C) 2022 Vaticle
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+define
+
+### Attributes ###
+
+    name sub attribute, value string;
+    date sub attribute, value datetime;
+    description sub attribute, value string;
+    id sub attribute, value string;
+
+    price sub attribute, value long;
+    stock sub attribute, value long;
+
+    ISBN sub attribute, value string;
+    book-author sub attribute, value string;
+    publisher sub attribute, value string;
+
+    rating sub attribute, value long;
+
+    age sub attribute, value long;
+
+    foreign-user-id sub attribute, value string;
+    status sub attribute, value string;
+    delivery-address sub attribute, value string;
+    payment-details sub attribute, value string;
+
+    username sub attribute, value string;
+    password sub attribute, value string;
+    foreign-id sub attribute, value string;
+
+    genre sub attribute, value string,
+        plays taging:tag,
+        plays tag-hierarchy:sup-tag,
+        plays tag-hierarchy:sub-tag;
+
+### Entities ###
+    product sub entity, abstract,
+        owns id @key,
+        owns name,
+        owns description,
+        owns price,
+        owns stock,
+        plays ordering:item,
+        plays reviewing:product;
+
+    book sub product,
+        owns ISBN,
+        owns book-author,
+        owns publisher,
+        plays taging:book;
+
+    person sub entity,
+        owns name,
+        plays reviewing:author;
+
+    user sub person,
+        owns id @key,
+        owns foreign-id,
+        owns username,
+        owns password,
+        owns age,
+        plays ordering:author;
+
+    order sub entity,
+        owns id @key,
+        owns foreign-user-id,
+        owns date,
+        owns status,
+        owns delivery-address,
+        owns payment-details,
+        plays ordering:order;
+
+### Relations ###
+    reviewing sub relation,
```

Comment:
You may have read or heard from various people that we like "verbal nouns" as our relation names: `employment`, `location`, `friendship` are all good examples.

In the case of `review` though, the simplest verbal noun is just `review` itself. We're representing a `review`, which exists because a `reviewer` reviewed a `product`.

Another option: `book-review`. This makes it entirely clear what is being reviewed (a `reviewer` reviewing a `book`), although it may be less flexible in the future should customers be able to leave reviews on products that are not books, such as stationery and board games. (And we wouldn't actually be able to call the role `book` because of name conflict - it'd have to be `reviewed-book`.)

Reply:
Fixed in https://github.com/vaticle/typedb-examples/pull/143/commits/4a1797d138a78fd029a1f1ec2c478090795f8e7c